File: isegoria/src/main/java/com/eulersbridge/isegoria/network/TimestampAdapter.java
package com.eulersbridge.isegoria.network;

import com.squareup.moshi.FromJson;
import com.squareup.moshi.ToJson;

import java.util.Calendar;
import java.util.TimeZone;

/**
 * Convert Unix timestamps in JSON to a Unix timestamp in the user's local timezone.
 */
@SuppressWarnings("unused")
class TimestampAdapter {

    private static int getUserOffsetFromUtc() {
        Calendar calendar = Calendar.getInstance();
        TimeZone timeZone = calendar.getTimeZone();
        return timeZone.getOffset(0) / 1000;
    }

    /**
     * Convert a timestamp in the user's timezone to UTC
     */
    @ToJson
    long toJson(@Timestamp long timestamp) {
        return timestamp - getUserOffsetFromUtc();
    }

    /**
     * Convert a JSON timestamp to one in the user's timezone
     */
    @FromJson
    @Timestamp long fromJson(long timestamp) {
        return timestamp + getUserOffsetFromUtc();
    }

}

File: isegoria/src/main/kotlin/com/eulersbridge/isegoria/network/adapters/TimestampAdapter.kt
package com.eulersbridge.isegoria.network.adapters

import com.squareup.moshi.FromJson
import com.squareup.moshi.ToJson
import java.util.*

/**
 * Convert Unix timestamps in JSON to a Unix timestamp in the user's local timezone.
 */
class TimestampAdapter {
    private val userOffsetFromUtc: Int
        get() {
            return Calendar.getInstance().timeZone.getOffset(0) / 1000
        }

    /**
     * Convert a timestamp in the user's timezone to UTC
     */
    @ToJson
    internal fun toJson(@Timestamp timestamp: Long)
            = timestamp - userOffsetFromUtc

    /**
     * Convert a JSON timestamp to one in the user's timezone
     */
    @FromJson
    @Timestamp
    internal fun fromJson(timestamp: Long)
            = timestamp + userOffsetFromUtc

}